### ISSUE TITLE
Use preemptible node for scans

### ISSUE THREAD

**Issue by l4mb0san** [secureCodeBox/secureCodeBox#803]
## Your Question

Currently, I am using preemptible nodes on my cluster to save costs, so I would like scans to be deployed on preemptible nodes. For secureCodeBox operator, I will deploy on a normal node (non-preemptible).

I tried apply 1 configuration as follows, but it doesn't work.
```
apiVersion: "execution.securecodebox.io/v1"
kind: Scan
metadata:
  name: "trivy-mediawiki"
spec:
  affinity:
    nodeAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
        nodeSelectorTerms:
        - matchExpressions:
          - key: cloud.google.com/gke-nodepool
            operator: In
            values:
            - preemptible
  tolerations:
  - key: "dedicated"
    operator: "Equal"
    value: "preemptible"
    effect: "NoSchedule" 
    
  scanType: "trivy"
  parameters:
    - "mediawiki:stable"
```
```
❯ k apply -f preemptible-trivy-test.yaml
error: error validating "preemptible-trivy-test.yaml": error validating data: [ValidationError(Scan.spec): unknown field "affinity" in io.securecodebox.execution.v1.Scan.spec, ValidationError(Scan.spec): unknown field "tolerations" in io.securecodebox.execution.v1.Scan.spec]; if you choose to ignore these errors, turn validation off with --validate=false
```

**So is there any way I can do this?**

**Comment by malexmave** [comment]
It looks like the `affinity` and `tolerations` fields are currently not explicitly exposed in the Kubernetes Custom Resource Definition (CRD) that we use for the scan specification. However, I don't see a reason why we shouldn't add them. While we're at it, are there any other fields in connection with this use case that you or other people might find useful to expose?

**Comment by l4mb0san** [comment]
> It looks like the `affinity` and `tolerations` fields are currently not explicitly exposed in the Kubernetes Custom Resource Definition (CRD) that we use for the scan specification. However, I don't see a reason why we shouldn't add them. While we're at it, are there any other fields in connection with this use case that you or other people might find useful to expose?

Maybe you misunderstanding sth, I said that we should add them.

My cluster is in common use for many different projects. If running multiple scans concurrently, it consumes a lot of resources, and my cluster used to be overloaded, which caused all the pods to be killed. It is for this reason that I want to specify a node to deploy (nodesSelector) and use preemptible node to save on deployment costs.

**Comment by EndPositive** [comment]
Hi @l4mb0san, I think the issues you're describing should be solved by defining the [resource requirements](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/) on your scanners. If you define this, Kubernetes will not schedule your scan unless it has enough resources to handle it.

The text below shows a truncated output for `kubectl describe nodes`. I've deployed the nmap scanner with request 35m cpu and 45Mi memory. This fits inside the free resources, so they are scheduled.

```
Non-terminated Pods:          (10 in total)
  Namespace                   Name                                                        CPU Requests  CPU Limits  Memory Requests  Memory Limits  Age
  ---------                   ----                                                        ------------  ----------  ---------------  -------------  ---
  securecodebox-scanners      scan-bv6d2qr-nmap-hostscan-tcp-intense-9mc8d-9d8c9-4dncn    35m (0%)      100m (2%)   45Mi (0%)        100Mi (1%)     8m43s
  securecodebox-scanners      scan-bv6d2qr-nmap-hostscan-tcp-intense-sn5r9-qrbw8-sc9lw    35m (0%)      100m (2%)   45Mi (0%)        100Mi (1%)     10h
  securecodebox-system        securecodebox-controller-manager-56cd9c486b-69jtk           100m (2%)     0 (0%)      20Mi (0%)        0 (0%)         22h
Allocated resources:
  (Total limits may be over 100 percent, i.e., overcommitted.)
  Resource           Requests     Limits
  --------           --------     ------
  cpu                622m (15%)   302m (7%)
  memory             715Mi (10%)  300Mi (4%)
```

You can define resources on scanners in their helm values:

```yaml
scanner:
  # scanner.resources -- CPU/memory resource requests/limits (see: https://kubernetes.io/docs/tasks/configure-pod-container/assign-memory-resource/, https://kubernetes.io/docs/tasks/configure-pod-container/assign-cpu-resource/)
  resources: {}
  #   resources:
  #     requests:
  #       memory: "256Mi"
  #       cpu: "250m"
  #     limits:
  #       memory: "512Mi"
  #       cpu: "500m"
```

**Comment by malexmave** [comment]
>Maybe you misunderstanding sth, I said that we should add them.
> 
> My cluster is in common use for many different projects. If running multiple scans concurrently, it consumes a lot of resources, and my cluster used to be overloaded, which caused all the pods to be killed. It is for this reason that I want to specify a node to deploy (nodesSelector) and use preemptible node to save on deployment costs.

Sorry, my "I don't see why we shouldn't add them" was actually meant in agreement with your point, meaning: I think it's a good idea to support this, so I am working on implementing it - didn't realize that the other reading was also valid :).

I agree with @EndPositive that defining resource requirements is a good practice, however, I think that your current issue is actually not related to this, but rather to the fact that you want to use a specific class of GCP nodes (preemptible nodes) that are cheaper, and which require adding the `tolerations` and `affinity` information to the scans. In general, I think that this is a very valid use case that should be supported - it would also allow other things in other clusters. For example, if you have multiple machines in your k8s cluster, some of which are running on-premise and others in the cloud, and you want to make sure that some specific scans are always run on-premise, then affinity and tolerations seem the way to go. Same for running scans from specific regions, etc.

So, short version: I'm in favor of supporting it and will try to get it into SCB for the next release. You can track my progress in #804.

**Comment by l4mb0san** [comment]
@malexmave Ya, thanks for your support ^^

@EndPositive  Yes, I'm also using it to limit resource usage now
